jetty-websocket/websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/util/MessageHandler.java

```diff
@@ -0,0 +1,175 @@
+                long size = _textMessageBuffer.length() + frame.getPayloadLength();
+                long maxTextMessageSize = getCoreSession().getMaxTextMessageSize();
+                if (maxTextMessageSize > 0 && size > maxTextMessageSize)
+                {
+                    throw new MessageTooLargeException(String.format("Text message too large: (actual) %,d > (configured max text message size) %,d",
+                        size, maxTextMessageSize));
+                }
+
+                _textMessageBuffer.append(frame.getPayload());
+            }
+
+            if (frame.isFin())
+            {
+                onText(_textMessageBuffer.toString(), callback);
+                _textMessageBuffer.reset();
+            }
+            else
+            {
+                callback.succeeded();
+            }
+        }
+        catch (Utf8Appendable.NotUtf8Exception e)
+        {
+            if (_textMessageBuffer != null)
+            {
+                _textMessageBuffer.reset();
+                _textMessageBuffer = null;
+            }
+            callback.failed(new BadPayloadException(e));
+        }
+        catch (Throwable t)
+        {
+            if (_textMessageBuffer != null)
+            {
+                _textMessageBuffer.reset();
+                _textMessageBuffer = null;
+            }
+            callback.failed(t);
+        }
+        finally
+        {
+            if (frame.isFin())
+                _textMessageBuffer = null;
+        }
+    }
+
+    @Override
+    protected void onBinaryFrame(Frame frame, Callback callback)
+    {
+        try
+        {
+            long size = (_binaryMessageBuffer == null ? 0 : _binaryMessageBuffer.size()) + frame.getPayloadLength();
+            long maxBinaryMessageSize = getCoreSession().getMaxBinaryMessageSize();
+            if (maxBinaryMessageSize > 0 && size > maxBinaryMessageSize)
+            {
+                throw new MessageTooLargeException(String.format("Binary message too large: (actual) %,d > (configured max binary message size) %,d",
+                    size, maxBinaryMessageSize));
+            }
+
+            // If we are fin and no OutputStream has been created we don't need to aggregate.
+            if (frame.isFin() && (_binaryMessageBuffer == null))
+            {
+                if (frame.hasPayload())
+                    onBinary(frame.getPayload(), callback);
+                else
+                    onBinary(BufferUtil.EMPTY_BUFFER, callback);
+                return;
+            }
+
+            // Aggregate the frame payload.
+            if (frame.hasPayload())
+            {
+                if (_binaryMessageBuffer == null)
+                    _binaryMessageBuffer = new ByteArrayOutputStream();
+                BufferUtil.writeTo(frame.getPayload(), _binaryMessageBuffer);
+            }
+
+            if (frame.isFin())
+            {
+                onBinary(BufferUtil.toBuffer(_binaryMessageBuffer.toByteArray()), callback);
```

Comment:
I feel the usage of `ByteArrayOutputStream` forces too many copies of the data.
Why you did not use `ByteBufferAccumulator`?

Reply:
It is not possible to use this with the `AutoDemandingMessageHandler`. We cannot receive the next message until the callback is succeeded and to succeed the callback means we no longer own the buffer, so we need to copy the buffer.

If you want more efficient binary message aggregation you should use the `DemandingMessageHandler`.

Reply:
`ByteBufferAccumulator` does exactly that, it copies the buffer, so you can use it.
Right now you're doing the same job as `ByteBufferAccumulator`, only using `ByteArrayOutputStream`.

Reply:
Yep agreed, I was thinking of `ByteBufferCallbackAccumulator`.